internal/engine/ooapi/clientcall_test.go

```diff
@@ -0,0 +1,898 @@
+	}
+	got := &apimodel.CheckInRequest{}
+	if err := json.Unmarshal(handler.body, &got); err != nil {
+		t.Fatal(err)
+	}
+	if diff := cmp.Diff(req, got); diff != "" {
+		t.Fatal(diff)
+	}
+}
+
+type handleClientCallMeasurementMeta struct {
+	accept      string
+	body        []byte
+	contentType string
+	count       int32
+	method      string
+	mu          sync.Mutex
+	resp        *apimodel.MeasurementMetaResponse
+	url         *url.URL
+	userAgent   string
+}
+
+func (h *handleClientCallMeasurementMeta) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ff := fakeFill{}
+	defer h.mu.Unlock()
+	h.mu.Lock()
+	if h.count > 0 {
+		w.WriteHeader(400)
+		return
+	}
+	h.count++
+	if r.Body != nil {
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(400)
+			return
+		}
+		h.body = data
+	}
+	h.method = r.Method
+	h.url = r.URL
+	h.accept = r.Header.Get("Accept")
+	h.contentType = r.Header.Get("Content-Type")
+	h.userAgent = r.Header.Get("User-Agent")
+	var out *apimodel.MeasurementMetaResponse
+	ff.fill(&out)
+	h.resp = out
+	data, err := json.Marshal(out)
+	if err != nil {
+		w.WriteHeader(400)
+		return
+	}
+	w.Write(data)
+}
+
+func TestMeasurementMetaClientCallRoundTrip(t *testing.T) {
+	// setup
+	handler := &handleClientCallMeasurementMeta{}
+	srvr := httptest.NewServer(handler)
+	defer srvr.Close()
+	req := &apimodel.MeasurementMetaRequest{}
+	ff := &fakeFill{}
+	ff.fill(&req)
+	clnt := &Client{KVStore: &MemKVStore{}, BaseURL: srvr.URL}
+	ff.fill(&clnt.UserAgent)
+	// issue request
+	ctx := context.Background()
+	resp, err := clnt.MeasurementMeta(ctx, req)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp == nil {
+		t.Fatal("expected non-nil response here")
+	}
+	// compare our response and server's one
+	if diff := cmp.Diff(handler.resp, resp); diff != "" {
+		t.Fatal(diff)
+	}
+	// check whether headers are OK
+	if handler.accept != "application/json" {
+		t.Fatal("invalid accept header")
+	}
+	if handler.userAgent != clnt.UserAgent {
+		t.Fatal("invalid user-agent header")
+	}
+	// check whether the method is OK
+	if handler.method != "GET" {
+		t.Fatal("invalid method")
+	}
+	// check the query
+	api := &simpleMeasurementMetaAPI{BaseURL: srvr.URL}
+	httpReq, err := api.newRequest(context.Background(), req)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if diff := cmp.Diff(handler.url.Path, httpReq.URL.Path); diff != "" {
+		t.Fatal(diff)
+	}
+	if diff := cmp.Diff(handler.url.RawQuery, httpReq.URL.RawQuery); diff != "" {
+		t.Fatal(diff)
+	}
+}
+
+type handleClientCallTestHelpers struct {
+	accept      string
+	body        []byte
+	contentType string
+	count       int32
+	method      string
+	mu          sync.Mutex
+	resp        apimodel.TestHelpersResponse
+	url         *url.URL
+	userAgent   string
+}
+
+func (h *handleClientCallTestHelpers) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ff := fakeFill{}
+	defer h.mu.Unlock()
+	h.mu.Lock()
+	if h.count > 0 {
+		w.WriteHeader(400)
+		return
+	}
+	h.count++
```

Comment:
TODO(@bassosimone): This should be atomic (not a big deal in this test but better to design correct types). There is similar code when testing the simple APIs, so we should also change that code.

Reply:
Ah, no, we already have a mutex.